### DISCUSSION TITLE
[FR] Publish vendor config and/or assets

### DISCUSSION THREAD

**Discussion by matfish2** [craftcms/cms#10161] [Ideas]
Laravel offers package developers the option to publish config and assets.
https://laravel.com/docs/8.x/packages#resources
I couldn't find an equivalent in Craft.

For the config file, this would save users the trouble of creating the file themselves from the docs, and will also serve as documentation for the plugin options.
For templates, images and CSS files it would allow a more standardized process where users can hoist plugin assets to the project folder and modify them.

Ideally users would be able to selectively decide which assets they want published (https://laravel.com/docs/8.x/packages#publishing-file-groups).

I apologize in advance if this functionality already exists. I couldn't find it.

**Comment by brandonkelly** [comment]
If I’m understanding you correctly, the Yii equivalent for publishing assets is “Asset Bundles” – see https://www.yiiframework.com/doc/guide/2.0/en/structure-assets.

As for config files, most plugins don’t require them and they are generally used for advanced configuration only, so I’m not sure it makes sense to automate their creation.

**Reply by matfish2** [reply]
AFAIU Asset Bundles allow read-only access to vendor assets by publishing the assets on runtime (if they don't already exist). E.g I see in Craft it  references `web/cpresources/{uuid}`. If I delete the folder it is regenerated on runtime. However, It does not allow modification. So if plugin has a `styles.css` file and user wants to modify it, he'll have to manually copy from the vendor folder. Case in point: I wrote a blog plugin and had to use vanilla PHP to copy the default templates to the project's `templates` folder. In Laravel I would be able to do it using the `publishes` method, and then either publish it on install or let the user run `php artisan vendor:publish` (https://laravel.com/docs/8.x/packages#publishing-views).  IMHO adding such functionality to Craft would make the plugin-project relationship more flexible when it comes to front-end resources. As for config files, in some cases there are developer-oriented settings that should not be exposed in the CP, or at least not worth the trouble of adding them to CP.

**Reply by brandonkelly** [reply]
Ahh gotcha.
